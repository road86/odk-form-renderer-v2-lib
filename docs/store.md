# Walkthrough through Redux store

## Store

The store is initialized in `src/store.tsx` and added in `src/index.tsx` file.

## Reducer

Currently, there is one single reducer, that resides in the `src/ducks/formstate.ts`.

### State of the store
The state of the store can be best represented as below:

```js
export interface FormState {
  userInput: object;
  errors: string[];
  optionList: object;
  isFormSubmitted: boolean;
}
```
### Brief description of the major keys

#### userInput

The `userInput` key contains the most updated form field values.

#### errors

The `errors` hold the list of field names that have input values missing when required or violating form constraints. If the form successfully passes all the constraints or does not have any required missing values, the `errors` key will be an empty array with no field names. And form submission will only be successful if this key is empty.

#### optionList 

The `optionList` holds all the options of single select and multiple select presents in the forms at the time of rendered. It add or removes any options based on user input of fields.

#### isFormSubmitted

The initial value of this key is `false`. It acts as a utility flag on whether to show required or constraint messages If a form gets submitted and the submission is not successful, the flag is assigned `true`.