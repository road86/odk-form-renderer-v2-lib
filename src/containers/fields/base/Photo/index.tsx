import * as React from 'react';
import { connect } from 'react-redux';
import { FormGroup, Input, Label } from 'reactstrap';
import { Store } from 'redux';
import { FieldElement } from '../../../../components/typeEvalutors/Base';
import { REQUIRED_FIELD_MSG, REQUIRED_SYMBOL } from '../../../../constants';
import {
  assignFieldValueAction,
  getFieldValue,
} from '../../../../store/ducks/formState';
import { getFieldLabelText, isInputRequired } from '../../../../utils/helpers';

/** props interface for the photo component */
export interface PhotoProps {
  fieldElement: FieldElement;
  fieldValue: string;
  assignFieldValueActionCreator: typeof assignFieldValueAction;
  defaultLanguage: string;
}

class Photo extends React.Component<PhotoProps> {
  public render() {
    const { fieldElement, fieldValue, defaultLanguage } = this.props;
    const isRequired = isInputRequired(fieldElement);
    const fieldLabel = getFieldLabelText(fieldElement, defaultLanguage);
    return (
      <FormGroup>
        <Label>{fieldLabel}</Label>
        {isRequired && <Label>{REQUIRED_SYMBOL}</Label>}
        <Input
          type="file"
          accept="image/*"
          name={fieldElement.name}
          onChange={this.onChangeHandler}
          value={fieldValue}
        />
        {isRequired && <Label>{REQUIRED_FIELD_MSG}</Label>}
      </FormGroup>
    );
  }
  /** sets the value of field element in store
   * @param {React.FormEvent<HTMLInputElement>} event - the onchange input event
   */
  private onChangeHandler = (event: React.FormEvent<HTMLInputElement>) => {
    this.props.assignFieldValueActionCreator(
      event.currentTarget.name,
      event.currentTarget.value
    );
  };
}

/** connect the component to the store */

/** Interface to describe props from mapStateToProps */
interface DispatchedStateProps {
  fieldValue: string;
}

/** Interface to describe props from parent */
interface ParentProps {
  fieldElement: FieldElement;
}

/** Map props to state  */
const mapStateToProps = (
  state: Partial<Store>,
  parentProps: ParentProps
): DispatchedStateProps => {
  const { fieldElement } = parentProps;
  const result = {
    fieldValue: getFieldValue(state, fieldElement.name),
  };
  return result;
};

/** map props to actions */
const mapDispatchToProps = {
  assignFieldValueActionCreator: assignFieldValueAction,
};

/** connect Photo component to the redux store */
const ConnectedPhoto = connect(
  mapStateToProps,
  mapDispatchToProps
)(Photo);

export default ConnectedPhoto;
