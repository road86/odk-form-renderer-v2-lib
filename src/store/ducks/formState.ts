import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';
import evaluater from '../../utils/compiler';
import {
  checkGroupedValuesForEmpty,
  emptyGroupedValues,
  getModifiedUserInputObject,
  getValueFromUserInputObj,
} from '../../utils/helpers';

/** interface for the store state */
export interface FormState {
  userInput: object;
  errors: string[];
  optionList: object;
}

// actions

/** FIELD_VALUE_ASSIGNED action type */
export const FIELD_VALUE_ASSIGNED = 'odk/reducer/form/FIELD_VALUE_ASSIGNED';
/** OPTION_LIST_ASSIGNED action type */
export const OPTION_LIST_ASSIGNED = 'odk/reducer/form/OPTION_LIST_ASSIGNED';
/** RESET_STORE action type */
export const RESET_STORE = 'odk/reducer/form/RESET_STORE';
/** ADD_ERROR_INPUT_ID action type */
export const ADD_ERROR_INPUT_ID = 'odk/reducer/form/ADD_ERROR_INPUT_ID';
/** REMOVE_ERROR_INPUT_ID action type */
export const REMOVE_ERROR_INPUT_ID = 'odk/reducer/form/REMOVE_ERROR_INPUT_ID';
/** EMPTY_GROUP_FIELDS */
export const EMPTY_GROUP_FIELDS = 'odk/reducer/form/EMPTY_GROUP_FIELDS';
/** REMOVE_GROUP_FIELDS_FROM_ERRORS */
export const REMOVE_GROUP_FIELDS_FROM_ERRORS =
  'odk/reducer/form/REMOVE_GROUP_FIELDS_FROM_ERRORS';
export const SET_USER_INPUT_OBJ = 'odk/reducer/form/SET_USER_INPUT_OBJ';

/** interface for ASSIGN_FIELD_VALUE action */
export interface AssignFieldValueAction extends AnyAction {
  fieldTreeName: string;
  fieldValue: any;
  type: typeof FIELD_VALUE_ASSIGNED;
}

/** interface for OPTION_LIST_ASSIGNED action */
export interface AssignOptionListAction extends AnyAction {
  fieldTreeName: string;
  fieldValue: any;
  type: typeof OPTION_LIST_ASSIGNED;
}

/** interface for RESET_STORE action */
export interface ResetStoreAction extends AnyAction {
  type: typeof RESET_STORE;
}

/** interface for ADD_ERROR_INPUT_ID action */
export interface AddErrorInputId extends AnyAction {
  fieldTreeName: string;
  type: typeof ADD_ERROR_INPUT_ID;
}

/** interface for REMOVE_ERROR_INPUT_ID action */
export interface RemoveErrorInputId extends AnyAction {
  fieldTreeName: string;
  type: typeof REMOVE_ERROR_INPUT_ID;
}

/** interface for EMPTY_GROUP_FIELDS action */
export interface EmptyGroupFields extends AnyAction {
  fieldTreeName: string;
  type: typeof EMPTY_GROUP_FIELDS;
}

/** interface for REMOVE_GROUP_FIELDS_FROM_ERRORS action */
export interface RemoveGroupFieldsFromErrors extends AnyAction {
  fieldTreeName: string;
  type: typeof REMOVE_GROUP_FIELDS_FROM_ERRORS;
}

/** interface for SET_USER_INPUT_OBJ action */
export interface SetUserInputObj extends AnyAction {
  userInputObj: any;
  type: typeof SET_USER_INPUT_OBJ;
}

/** Assigns the value to the proper field name
 * @param {string} fieldTreeName - the extended field name
 * @param {any} fieldValue - the value that will be assigned
 * @return {AssignFieldValueAction} - an action to assign value to a field in the redux store
 */
export const assignFieldValueAction = (
  fieldTreeName: string,
  fieldValue: any
): AssignFieldValueAction => ({
  fieldTreeName,
  fieldValue,
  type: FIELD_VALUE_ASSIGNED,
});

/** Assigns option list to the proper field name
 * @param {string} fieldTreeName - the extended field name
 * @param {any} fieldValue - the option list that will be assigned
 * @return {AssignOptionListAction} - an action to assign option List to a field in the redux store
 */
export const assignOptionListAction = (
  fieldTreeName: string,
  fieldValue: any
): AssignOptionListAction => ({
  fieldTreeName,
  fieldValue,
  type: OPTION_LIST_ASSIGNED,
});

/** Resets the redux store state to initial state
 * @return {ResetStoreAction} - an action to reset the redux store state
 */
export const resetStoreAction = (): ResetStoreAction => ({
  type: RESET_STORE,
});

/** add the field tree name as error id to store in redux store that violates constraints
 * @param fieldTreeName - the field tree name
 * @returns {AddErrorInputId} - an action to add input id for errors
 */
export const addErrorInputId = (fieldTreeName: string): AddErrorInputId => ({
  fieldTreeName,
  type: ADD_ERROR_INPUT_ID,
});

/** remove the field tree name as error id to store in redux store that violates constraints
 * @param fieldTreeName - the field tree name
 * @returns {RemoveErrorInputId} - an action to remove input id for errors
 */
export const removeErrorInputId = (
  fieldTreeName: string
): RemoveErrorInputId => ({
  fieldTreeName,
  type: REMOVE_ERROR_INPUT_ID,
});

/** empties the values of the fields of the group
 * @param fieldTreeName - the group field tree name
 * @returns {RemoveErrorInputId} - an action to empty the group field values
 */
export const emptyGroupFields = (fieldTreeName: string): EmptyGroupFields => ({
  fieldTreeName,
  type: EMPTY_GROUP_FIELDS,
});

/** removes the group field names from store errors obj
 * @param {string} fieldTreeName - the group field tree name
 * @returns {RemoveGroupFieldsFromErrors} - an action to remove group field names from errors
 */
export const removeGroupFieldsFromErrors = (
  fieldTreeName: string
): RemoveGroupFieldsFromErrors => ({
  fieldTreeName,
  type: REMOVE_GROUP_FIELDS_FROM_ERRORS,
});

/** sets the user input object to redux store
 * @param {any} userInputObj - the user input obj
 * @returns {SetUserInputObj} - an action to set user input to redux store
 */
export const setUserInputObj = (userInputObj: any): SetUserInputObj => ({
  type: SET_USER_INPUT_OBJ,
  userInputObj,
});

/** Create type for forms reducer actions */
export type FormActionTypes =
  | AssignFieldValueAction
  | AssignOptionListAction
  | ResetStoreAction
  | AddErrorInputId
  | RemoveErrorInputId
  | EmptyGroupFields
  | RemoveGroupFieldsFromErrors
  | SetUserInputObj
  | AnyAction;

/** Create an immutable form state */
export type ImmutableFormState = SeamlessImmutable.ImmutableObject<FormState>;

/** initial form state */
export const initialState: ImmutableFormState = SeamlessImmutable({
  errors: [],
  optionList: {},
  userInput: {},
});

/** the form reducer function */
export default function reducer(
  state: ImmutableFormState = initialState,
  action: FormActionTypes
): ImmutableFormState {
  switch (action.type) {
    case FIELD_VALUE_ASSIGNED:
      const modifiedUserInputObj = getModifiedUserInputObject(
        state.getIn(['userInput']).asMutable({ deep: true }),
        action.fieldTreeName,
        action.fieldValue
      );
      const stateM = state.asMutable({ deep: true });
      return SeamlessImmutable({ ...stateM, userInput: modifiedUserInputObj });
    case OPTION_LIST_ASSIGNED:
      const modifiedUserInputObjList = getModifiedUserInputObject(
        state.getIn(['optionList']).asMutable({ deep: true }),
        action.fieldTreeName,
        { ...action.fieldValue }
      );
      const newState = state.asMutable({ deep: true });
      return SeamlessImmutable({
        ...newState,
        optionList: modifiedUserInputObjList,
      });
    case RESET_STORE:
      return initialState;
    case ADD_ERROR_INPUT_ID:
      if (!state.errors.includes(action.fieldTreeName)) {
        return state.updateIn(['errors'], arr =>
          arr.concat([action.fieldTreeName])
        );
      }
      return state;
    case REMOVE_ERROR_INPUT_ID:
      if (state.errors.includes(action.fieldTreeName)) {
        return state.updateIn(['errors'], arr =>
          arr.filter(elm => elm !== action.fieldTreeName)
        );
      }
      return state;
    case EMPTY_GROUP_FIELDS:
      const mUserInputObj = emptyGroupedValues(
        state.getIn(['userInput']).asMutable({ deep: true }),
        action.fieldTreeName
      );
      const mState = state.asMutable({ deep: true });
      return SeamlessImmutable({ ...mState, userInput: mUserInputObj });
    case REMOVE_GROUP_FIELDS_FROM_ERRORS:
      return state.updateIn(['errors'], arr =>
        arr.filter(elm => !elm.startsWith(action.fieldTreeName))
      );
    case SET_USER_INPUT_OBJ:
      return SeamlessImmutable({
        ...state,
        userInput: (action as any).userInputObj,
      });
    default:
      return state;
  }
}

// selectors

/** get the value by their respective element tree name
 * @param {Partial<Store>} state - the redux store
 * @param {string} fieldTreeName - the hierchical tree name of the field
 * @return {any | null} value if the element name is found else null
 */
export function getFieldValue(
  state: Partial<Store>,
  fieldTreeName: string
): any {
  return getValueFromUserInputObj((state as any).userInput, fieldTreeName);
}

/** get option list by their respective element tree name
 * @param {Partial<Store>} state - the redux store
 * @param {string} fieldTreeName - the hierchical tree name of the field
 * @return {any | null} value if the element name is found else null
 */
export function getOptionList(
  state: Partial<Store>,
  fieldTreeName: string
): any {
  return getValueFromUserInputObj((state as any).optionList, fieldTreeName);
}

/** get the value of the evaluated expression
 * @param {Partial<Store>} state - the redux store
 * @param {string} expression - the expression that needs to be evaluated
 * @param {string} fieldTreeName - the hierchical tree name of the field
 * @return {any | null} - the evaluated value or null if syntax error
 */
export function getEvaluatedExpression(
  state: Partial<Store>,
  expression: string,
  fieldTreeName: string
): any {
  return evaluater(expression, (state as any).userInput, null, fieldTreeName);
}

/** get the value of the evaluated expression for Select One and Select All
 * @param {Partial<Store>} state - the redux store
 * @param {string} expression - the expression that needs to be evaluated
 * @param {any} options - the options that needs to be evaluated
 * @param {string} fieldTreeName - the hierchical tree name of the field
 * @return {any | null} - the evaluated value or null if syntax error
 */
export function getEvaluatedExpressionForSelect(
  state: Partial<Store>,
  expression: string,
  options: any,
  fieldTreeName: string
): any {
  return evaluater(
    expression,
    (state as any).userInput,
    options,
    fieldTreeName
  );
}

/** check if the field element is present in the store
 * @param {Partial<Store>} state - the redux store
 * @param {string} fieldTreeName - the hierchical tree name of the field
 * @return {boolean} true if present; otherwise, false
 */
export function isPresentInError(
  state: Partial<Store>,
  fieldTreeName: string
): any {
  return (state as any).errors.includes(fieldTreeName);
}

/** check if the field elements under group are empty or not
 * @param {Partial<Store>} state - the redux store
 * @param {string} fieldTreeName - the hierchical tree name of the group field
 * @return {boolean} true if empty; otherwise, false
 */
export function isGroupFieldsEmpty(
  state: Partial<Store>,
  fieldTreeName: string
): any {
  return checkGroupedValuesForEmpty((state as any).userInput, fieldTreeName);
}

/** check if the field elements under group are present in errors or not
 * @param {Partial<Store>} state - the redux store
 * @param {string} fieldTreeName - the hierchical tree name of the group field
 * @return {boolean} true if present; otherwise, false
 */
export function isErrorsIncludeGroupFields(
  state: Partial<Store>,
  fieldTreeName: string
): any {
  let isPresent = false;
  (state as any).errors.forEach((fTName: string) => {
    if (fTName.startsWith(fieldTreeName) && !isPresent) {
      isPresent = true;
    }
  });
  return isPresent;
}

/** check if the errors array in store empty or not
 * @param {Partial<Store>} state - the redux store
 * @return {boolean} true if empty; otherwise, false
 */
export function isErrorsArrayEmpty(state: Partial<Store>): any {
  return (state as any).errors && (state as any).errors.length ? false : true;
}

/** get the userInput object from store
 * @param {Partial<Store>} state - the redux store
 * @return {boolean} the current userInputObject
 */
export function getUserInputFromStore(state: Partial<Store>): any {
  return (state as any).getIn(['userInput']).asMutable({ deep: true });
}
