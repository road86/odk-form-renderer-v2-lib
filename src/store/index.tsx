import { createStore, Store } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import reducer, { FormState } from './ducks/formState';

/** The initial store */
const store: Store & FormState = createStore(reducer, composeWithDevTools());

export default store;
