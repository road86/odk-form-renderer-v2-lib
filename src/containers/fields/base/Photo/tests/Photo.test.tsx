import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { Provider } from 'react-redux';
import Photo from '..';
import store from '../../../../../store';
import { assignFieldValueAction } from '../../../../../store/ducks/formState';
import { photoFieldElement } from './fixtures';

describe('containers/fields/base/Photo', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    const mock: any = jest.fn();
    const props = {
      assignFieldValueActionCreator: mock,
      defaultLanguage: 'English',
      fieldElement: photoFieldElement,
      fieldValue: '',
    };
    shallow(
      <Provider store={store}>
        <Photo {...props} />
      </Provider>
    );
  });

  it('renders correctly', () => {
    const mock: any = jest.fn();
    const props = {
      assignFieldValueActionCreator: mock,
      defaultLanguage: 'English',
      fieldElement: photoFieldElement,
      fieldValue: '',
    };
    store.dispatch(
      assignFieldValueAction(props.fieldElement.name, props.fieldValue)
    );
    const wrapper = mount(
      <Provider store={store}>
        <Photo {...props} />
      </Provider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
