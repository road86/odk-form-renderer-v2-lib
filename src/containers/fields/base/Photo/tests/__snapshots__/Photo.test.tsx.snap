// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/fields/base/Photo renders correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(Photo)
    assignFieldValueActionCreator={[MockFunction]}
    defaultLanguage="English"
    fieldElement={
      Object {
        "hint": "image type with no appearance",
        "label": "Image widget",
        "name": "image_widget",
        "type": "photo",
      }
    }
    fieldValue=""
  >
    <Photo
      assignFieldValueActionCreator={[Function]}
      defaultLanguage="English"
      fieldElement={
        Object {
          "hint": "image type with no appearance",
          "label": "Image widget",
          "name": "image_widget",
          "type": "photo",
        }
      }
      fieldValue=""
    >
      <FormGroup
        tag="div"
      >
        <div
          className="form-group"
        >
          <Label
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <label
              className=""
            >
              Image widget
            </label>
          </Label>
          <Input
            accept="image/*"
            name="image_widget"
            onChange={[Function]}
            type="file"
            value=""
          >
            <input
              accept="image/*"
              className="form-control-file"
              name="image_widget"
              onChange={[Function]}
              type="file"
              value=""
            />
          </Input>
        </div>
      </FormGroup>
    </Photo>
  </Connect(Photo)>
</Provider>
`;
